### DISCUSSION TITLE
Run HFS when Windows starts; Minimize to tray clicking the close button [X]

### DISCUSSION THREAD

**Discussion by mitobin** [rejetto/hfs#282] [Q&A]
Can add Run HFS when Windows starts feature; Minimize to tray clicking the close button [X] from HFS2 to HFS3?
![HFS PNG](https://github.com/rejetto/hfs/assets/132363335/53caaa4f-5743-47bc-b1ee-5cfb7a1f5bfa)

**Comment by rejetto** [comment]
at the moment HFS 3 doesn't have those features.

run at start: you can find how to start any program when your Windows start.
The simple way, also used by HFS 2, is that the program starts after user logon.
Some people wants their server to start even without logon, in case the computer reboots and is unattended. This way requires the software to be configured as "service" in the system. You can read more about that if you are interested.

minimize: at the moment HFS 3 has no graphical "window", just a text window (also known as terminal or console). Text windows "go to tray". I planned (long ago) to have a window, at least for Windows users. To make this possible, a separate program must be done. It will arrive before or later. It would be cool if someone decided to help with this, as I'm busy with other features.

**Reply by mitobin** [reply]
Many thanks for taking the time to assist.

**Reply by f2065** [reply]
> To make this possible, a separate program must be done. It will arrive before or later. It would be cool if someone decided to help with this, as I'm busy with other features.

I made a small program for the tray, but there is a difficulty with finding the console window for control.
Especially universal for Windows 8.1, Windows 10, Windows 11.
The console window is created differently everywhere!
For Windows 8.1 and Windows 10 22H2:
The console window is a child window of the HFS3.exe process – having found the process windows, it is easy to find the window of the desired console. Although this is a slow method – you need to go through all the windows (with the class "ConsoleWindowClass" in the system, checking which exe file they are associated with.
For Windows 11 23H2:
The console window is not a window of the HFS3 process, there is a completely different logic of the console window.
Now I am looking for a window with the class "CASCADIA_HOSTING_WINDOW_CLASS" and the name "Windows PowerShell".
Then I check some indirect factors of the found process, but this is a bad algorithm, unreliable.
Is it possible to make some improvements from the HFS3 side, for easy search of the console by third-party programs?
For example, make a more unique name for the console window - "Rejetto Http File Server v3 Console".
Or some other methods that allow using WinAPI to easily find out the hWnd of the window of this console.

**Reply by rejetto** [reply]
hey! with node i doubt i have much power over it, but i can consider integrating [this C module](https://www.npmjs.com/package/node-hide-console-window) and exposing the commands somehow.
Well, it would be probably a good idea if your program was distributed as a plugin. Once i make the plugin i just said you could extend it, by putting your exe file inside and launch it.
If you are in for it, i think we should start a separate thread/discussion for the development.
Let me know.

**Reply by f2065** [reply]
> with node i doubt i have much power over it,

Is it possible to attach some other console?
In Windows 11 23H2, the overly complex powershell console works here.
It is not associated with the parent HFS process. It may contain windows from other processes.
As a result, the console connected to HFS is annoying and should be hidden, but in this same console, tabs from other processes may appear that do not need to be hidden. Problem.
In Windows 10, everything is simple - there the console is associated with the parent process, and there are no outsiders. But in Windows 11 22H3, everything is different.

**Reply by rejetto** [reply]
> Is it possible to attach some other console?

i'm not sure what you mean but, as i said, hfs may use that module and you can ask hfs to do the hiding

**Reply by f2065** [reply]
> hfs may use that module and you can ask hfs to do the hiding

Let's try...
Although I'm not sure it will work in Windows 11...
How will it be possible to control hiding the console from external processes?

**Reply by rejetto** [reply]
sorry for the later reply.
this deserves a dedicated discussion, as i said, so let's move to #830
